### DISCUSSION TITLE
issue: defaultValue of `false` will always return undefined for checkboxes

### DISCUSSION THREAD

**Discussion by mmahalwy** [react-hook-form/react-hook-form#7341] [Q&A] [answered]
### Version Number

7.22.2

### Codesandbox/Expo snack

https://codesandbox.io/s/condescending-rubin-8s3qc?file=/src/App.tsx

### Steps to reproduce

1. Go to sandbox link
2. Wait for 2s
3. Click submit
4. Notice the console log, the `user.approve` will be undefined.

I think this is due to a problem here: https://github.com/react-hook-form/react-hook-form/blob/master/src/utils/get.ts#L15. It checks if it's an object. When there is a nested field which is not defined, it will skip the `if` statement and return `undefined`. I think this function should return `defaultValue` if present, or `undefined`. Happy to create a PR for this. 

Technically, there is a `defaultValue` of false but is being ignored. I know that we shouldnt be setting `undefined` but this is a nested object, making it very difficult to control, especially when coming from API. 

### Expected behaviour

Should have default value of `false` honored. 

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by JamesWong1999** [comment, marked as answer]
https://react-hook-form.com/api/usecontroller

according to the doc it's either

> You need to either set defaultValue at the field-level or call useForm with defaultValues. If both are set,devaultValues will be used.

**Reply by hines-sharrod** [reply]
How do you set defaultValue on nested items in a field array that are dynamic and not set from the start? 

I am having a similar problem when I setup my form and try to fill in other sections and not my checkbox I am expecting it to default to false naturally. I am using a fieldArray so there is no way to set defaultValues using useForm or on the field level for dynamic data, so the checkbox keeps being stored as undefined instead of false.

**Comment by bluebill1049** [comment]
thanks, @JamesWong1999 that's correct. it's either.

**Comment by mmahalwy** [comment]
@bluebill1049 @JamesWong1999 got it, thoughts on how to make this work when default values are coming from async call and now all values are set? I cant imagine replacing `defaultValues` on `useForm` to  `defaultValue: apiResponse.some.path.to.value` on every field. This is especially true for dynamic fields (hide/show) where `defaultValues` is already set.

**Comment by mmahalwy** [comment]
Feels like we should either:
1. have an option on `useController` to allow for `defaultValue` to override
2. If `undefined` on `defaultValues`, prefer the `defaultValue` on `useController`

Thoughts?

**Comment by bluebill1049** [comment]
please use the following APIs

- `setValue`
- `reset`

**Comment by mmahalwy** [comment]
I'm using reset in the example given in codesandbox 😔

I also don't think I can do setValue on 10s of fields

**Comment by bluebill1049** [comment]
![Screen Shot 2021-12-22 at 7 39 02 pm](https://user-images.githubusercontent.com/10513364/147062489-136e3758-324a-4ea3-88db-38343ba3b4e3.png)

I don't see `reset`.

**Comment by mmahalwy** [comment]
@bluebill1049 sorry, forgot to save codesandbox. Here: https://codesandbox.io/s/condescending-rubin-8s3qc?file=/src/App.tsx


![image](https://user-images.githubusercontent.com/3749095/147063215-8084b654-352b-415c-85cf-56f3d1fcc499.png)

**Comment by bluebill1049** [comment]
```tsx
import * as React from "react";
import { useController, useForm } from "react-hook-form";
import Headers from "./Header";
import "./styles.css";

type FormValues = {
  firstName: string;
};

let renderCount = 0;

const Checkbox = ({ control }) => {
  const { field } = useController({
    control,
    name: "user.approved"
  });

  return (
    <label>
      <input {...field} type="checkbox" /> Checkbox
    </label>
  );
};

export default function App() {
  const { control, handleSubmit, reset } = useForm<FormValues>({
    defaultValues: {
      user: {
        approved: false,
      }
    }
  });

  React.useEffect(() => {
    setTimeout(() => {
      reset({
        user: {
          approved: false,
        }
      });
    }, 1000);
  }, [reset]);

  const onSubmit = (data: FormValues) => console.log(data);
  renderCount++;

  return (
    <div>
      <Headers
        renderCount={renderCount}
        description="Performant, flexible and extensible forms with easy-to-use validation."
      />

      <form onSubmit={handleSubmit(onSubmit)}>
        <Checkbox control={control} />
        <input type="submit" />
      </form>
    </div>
  );
}
```

Do above.

**Comment by mmahalwy** [comment]
hey @bluebill1049 this wouldnt exactly solve my problem as the default values are coming from API, hence why the `useEffect` to show that. The shape is not guaranteed to include `{ user: { approve: false } }`. API can return:
```
{ id: 123, user: undefined }
```
which is common in GraphQL applications where `user` is nullable.

**Comment by mmahalwy** [comment]
Here's an updated example that doesnt have nesting but still the same problem: https://codesandbox.io/s/condescending-rubin-8s3qc?file=/src/App.tsx. API does not return `approved` but the form has default. Merging the two would be very challenging in `defaultValues`.

**Reply by bluebill1049** [reply]
```
const { field } = useController({
  defaultValue: false,
  control,
  name: "approved"
});
```

why persist in putting `defaultValue` at `useController`? 

`{ id: 123, user: undefined }` if this is the case, the issue should get taken care the component level, hook form will remain either for default values or defaultValue, which I would recommend using `defautlValues`?

**Reply by mmahalwy** [reply]
Honestly, I had to go back to final-form because this was such a common use case for us. There should be a way to overwrite the default values on the `useController` level if needed.
